smartsim/_core/mli/infrastructure/control/devicemanager.py

```diff
@@ -0,0 +1,130 @@
+        self._lock = RLock()
+        """Lock to ensure only one thread at the time accesses this device"""
+        self._models: dict[str, t.Any] = {}
+
+    def acquire(self, blocking: bool = True, timeout: float = -1) -> t.Optional[bool]:
+        return self._lock.acquire(blocking=blocking, timeout=timeout)
+
+    def release(self) -> None:
+        self._lock.release()
+
+    def __enter__(self) -> None:
+        self.acquire()
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    def add_model(self, key: str, model: t.Any) -> None:
+        self._models[key] = model
+
+    def remove_model(self, key: str) -> None:
+        self._models.pop(key)
+
+    def get_model(self, key: str) -> t.Any:
+        return self._models[key]
+
+    def __contains__(self, key: str) -> bool:
+        return key in self._models
+
+    def __exit__(
+        self,
+        exc_type: t.Optional[t.Type[BaseException]],
+        exc_val: t.Optional[BaseException],
+        exc_tb: t.Optional[TracebackType],
+    ) -> None:
+        self.release()
+
+
+class DeviceManager:
+    def __init__(self, devices: list[WorkerDevice]):
+        self._devices = devices
+        """Dictionary of model key to devices on which it is loaded"""
+
+    def get_free_device(
+        self,
+        worker: MachineLearningWorkerBase,
+        batch: InferenceBatch,
+        feature_store: t.Optional[FeatureStore],
+    ) -> t.Generator[WorkerDevice, None, None]:
+        return_device = None
+        sample_request = batch.requests[0]
+        direct_inference = sample_request.raw_model is not None
+        while return_device is None:
+            loaded_devices = []
+            if not direct_inference:
+                # Look up devices to see if any of them already has a copy of the model
+                for device in self._devices:
+                    if batch.model_key in device:
+                        loaded_devices.append(device)
+
+                # If a pre-loaded model is found on a device, try using that device
+                for device in loaded_devices:
+                    if device.acquire(blocking=False):
+                        return_device = device
+                        break
+
+            # If the model is not loaded on a free device,
+            # load it on another device (if available)
+            if return_device is None:
+                for candidate_device in self._devices:
+                    if (
+                        candidate_device not in loaded_devices
+                        and candidate_device.acquire(blocking=False)
+                    ):
+                        model_bytes = worker.fetch_model(batch, feature_store)
+                        loaded_model = worker.load_model(
+                            batch, model_bytes, candidate_device.name
+                        )
+                        candidate_device.add_model(batch.model_key, loaded_model.model)
```

Comment:
Might want to catch any exceptions that could occur in `fetch_model` and `load_model`

Reply:
Right, I'm going to do it after I remove all the multi-threading logic.